Add devenv (#807)
shopware/docs: guides/installation/devenv.md

```shell
sh <(curl -L https://nixos.org/nix/install) --daemon
```

{% endtab %}

{% tab title="Windows (WSL2)" %}

```shell
sh <(curl -L https://nixos.org/nix/install) --no-daemon
```

{% endtab %}

{% tab title="Docker" %}

```shell
docker run -it nixos/nix
```

{% endtab %}
{% endtabs %}

### Cachix

Next, install [Cachix](https://www.cachix.org/) to speed up the installation:

```shell
nix-env -iA cachix -f https://cachix.org/api/v1/install
```

Before installing devenv, instruct Cachix to use the devenv cache:

```shell
cachix use devenv
```

{% hint style="info" %}
When running `cachix use ...` for the first time, you will see a warning that your user is not in trusted-users.

```bash
This user doesn't have permissions to configure binary caches.

You can either:
...
```

When faced with this message, you can run

```shell
echo "trusted-users = root ${USER}" | sudo tee -a /etc/nix/nix.conf && sudo pkill nix-daemon
```

{% endhint %}

### Devenv

Finally, install devenv:

```shell
nix-env -if https://github.com/cachix/devenv/tarball/v0.5
```

Before booting up your development environment, configure Cachix to use Shopware's cache:

```shell
cachix use shopware
```

> By default, `shopware/platform` uses unfree software like Blackfire.
> To be able to use unfree software, you have to allow that:

Reviewer: ```suggestion
By default, `shopware/platform` uses unfree software like Blackfire. To be able to use unfree software, you have to *allow* it:
```

```bash
mkdir -p ~/.config/nixpkgs
echo '{ allowUnfree = true; }' > ~/.config/nixpkgs/config.nix
```

You can find the whole installation guide for devenv in their official documentation:

<!-- markdown-link-check-disable-next-line -->
{% embed url="https://devenv.sh/getting-started/" caption="Getting started - devenv.sh" %}

### Shopware

Depending on whether you want to set up a fresh Shopware project or contribute to the Shopware core, choose between:

{% tabs %}
{% tab title="Symfony Flex" %}

> If you are already using Symfony Flex, you can require a Composer package to get a basic devenv configuration:

Reviewer: ```suggestion
If you are already using Symfony Flex, you require a composer package to get a basic devenv configuration:
```

```bash
composer require devenv
```

This will create a basic `devenv.nix` file to enable devenv support for Shopware.
{% endtab %}

{% tab title="shopware/platform (Contribute)" %}
Clone [shopware/platform](https://github.com/shopware/platform) and change into the project directory:

```shell
git clone git@github.com:shopware/platform.git
```

{% endtab %}
{% endtabs %}

Since the environment is described via a `devenv.nix` file committed to version control, you can now boot up the environment:

```shell
devenv up
```

> {% hint style="info" %}
> Make sure that the ports for the services are not already in use or the command will fail.
> {% endhint %}

Reviewer: ```suggestion
{% hint style="warning" %}
Make sure that the ports for the services are not already in use, or else the command will fail.
{% endhint %}
```

Ensure to change your `.env` file to have the database connect using localhost's IP address instead of the default MySQL socket:

{% code title="<PROJECT_ROOT>/.env" %}

```dotenv
DATABASE_URL="mysql://shopware:shopware@127.0.0.1:3306/shopware?sslmode=disable&charset=utf8mb4"
```

{% endcode %}

> Still inside the project directory, run the following command to initialize Shopware:

Reviewer: ```suggestion
Inside the project directory, run the following command to initialize Shopware:
```

```shell
composer setup
```

### Direnv

> If you wish to seamlessly switch between multiple development environments which use devenv we recommend to install [direnv](https://direnv.net/).

Reviewer: ```suggestion
If you wish to switch between multiple development environments which use devenv seamlessly, we recommend installing [direnv](https://direnv.net/).
```

```bash
apt install direnv
```

{% endtab %}

{% tab title="Other" %}
The installation instructions for other OS are available on direnv's [official documentation](https://direnv.net/docs/hook.html).
{% endtab %}
{% endtabs %}

Afterwards, add the following hook to your shell:

{% tabs %}

{% tab title="Bash" %}
{% code title="~/.bashrc" %}

```bash
eval "$(direnv hook bash)"
```

{% endcode %}
{% endtab %}

{% tab title="Zsh" %}
{% code title="~/.zshrc" %}

```bash
eval "$(direnv hook zsh)"
```

{% endcode %}
{% endtab %}

{% tab title="Fish" %}
{% code title="~/.config/fish/config.fish" %}

```bash
>direnv hook fish | source
```

{% endcode %}
{% endtab %}

{% tab title="Other" %}
The installation instructions for other OS are available on direnv's [official documentation](https://direnv.net/docs/hook.html).
{% endtab %}

{% endtabs %}

After you change into a project directory using devenv for the first time, you need to allow direnv to load the environment:

```bash
direnv allow
```

<!-- markdown-link-check-disable-next-line -->
{% embed url="https://devenv.sh/automatic-shell-activation/" caption="Automatic Shell Activation - devenv.sh" %}

## Default services

> Here's an overview of services Shopware provides by default and how you can access them:

Reviewer: ```suggestion
Here is an overview of services Shopware provides by default and how you can access them:
```

| Service          | Access                                         |
|------------------|------------------------------------------------|
| MySQL            | `mysql://shopware:shopware@127.0.0.1:3306`     |
| Caddy            | [http://localhost:8000](http://localhost:8000) |
| Adminer          | [http://localhost:9080](http://localhost:9080) |
| Mailhog (SMTP)   | `smtp://127.0.0.1:1025`                        |
| Mailhog (Web UI) | [http://localhost:8025](http://localhost:8025) |

## Customize your setup

> To customize the predefined services to match your needs, e.g. changing the virtual host, database name or environment variables, you can create `devenv.local.nix` to override the service definitions.

Reviewer: ```suggestion
To customize the predefined services to match your needs, e.g., changing the virtual host, database name, or environment variables, you can create `devenv.local.nix` to override the service definitions.
```

    extraConfig = ''
      # Copy the config from devenv.nix and append the XDebug config
      # [...]
      xdebug.mode=debug
      xdebug.discover_client_host=1
      xdebug.client_host=127.0.0.1
    '';
  };
}
```

{% endcode %}

## Known issues

### Manually reloading direnv

If you decided against using direnv, keep in mind that on every change to the `*.nix` files you need to manually reload the environment:

```shell
direnv reload
```

### Direnv slow in big projects

The bigger your project directory is getting over time (e.g. cache files piling up), the slower direnv will be.
This is a known issue and the devenv developers are working on a solution.

<!-- markdown-link-check-disable-next-line -->
{% embed url="https://github.com/cachix/devenv/issues/257" caption="Devenv slows down with big code repositories #257" %}

## FAQ

### How do I clean up devenv?

Periodically run `devenv gc` to remove orphaned services, packages and processes and free up disk space.

### How do I access the database?

The MySQL service is exposed under its default port `3306`, see [Default services](#default-services).

> Be aware that you cannot connect using the `localhost` socket, you must use `127.0.0.1` instead.

Reviewer: ```suggestion
Be aware that you cannot connect using the `localhost` socket. Instead, you must use `127.0.0.1`.
```
